[WIP] prototype bounded staleness end to end

otan opened the pull request.

nvb committed:
kv: implement server-side negotiation fast-path and Txn.NegotiateAndSend

Fixes #67551.
Fixes #67552.
Fixes #67553.
Touches #67562.

Bounded-staleness read orchestration consists of two phases -
negotiation and execution. Negotiation determines the timestamp to run
the query at in order to ensure that the read will not block on
replication or on conflicting transactions. Execution then uses this
timestamp to run the read request.

This commit implements the bounded staleness server-side negotiation
fast-path. This fast-path allows a bounded staleness read request that
lands on a single range to perform its negotiation phase and execution
phase in a single RPC.

The server-side negotiation fast-path provides two benefits:
1. it avoids two network hops in the common-case where a bounded
   staleness read is targeting a single range. This in an important
   performance optimization for single-row point lookups.
2. it provides stronger guarantees around minimizing staleness during
   bounded staleness reads. Bounded staleness reads that hit the
   server-side fast-path use their target replica's most up-to-date
   resolved timestamp, so they are as fresh as possible. Bounded
   staleness reads that miss the fast-path and perform explicit
   negotiation (see below) consult a cache, so they may use an
   out-of-date, suboptimal resolved timestamp, as long as it is fresh
   enough to satisfy the staleness bound of the request.

The commit then uses this new functionality to implement the
`(*Txn).NegotiateAndSend` method detailed in the bounded staleness RFC.
`NegotiateAndSend` is a specialized version of `Send` that is capable of
orchestrating a bounded-staleness read through a transaction, given a
read-only BatchRequest with a `min_timestamp_bound` set in its Header.
If the method returns successfully, the transaction will have been given
a fixed timestamp equal to the timestamp that the read-only request was
evaluated at.

nvb committed:
kv: add strict version of bounded staleness reads

This commit adds a `min_timestamp_bound_strict` flag to the BatchRequest
header. This flag specifies whether a bounded staleness read whose
min_timestamp_bound cannot be satisfied by the first replica it visits
(subject to routing_policy) without blocking should be rejected with a
MinTimestampBoundUnsatisfiableError or will be redirected to the
leaseholder and permitted to block on conflicting transactions. If the
flag is true, blocking is never permitted and users should be prepared
to handle MinTimestampBoundUnsatisfiableErrors. If the flag is false,
blocking is permitted and MinTimestampBoundUnsatisfiableErrors will
never be returned.

This will be used by SQL to implement the `nearest_only` bounded
staleness option described in the RFC. It will also be used to implement
the descriptor iteration logic that steps back across table descriptor
versions and only ever blocks on the earliest descriptor.

nvb committed:
kv: add optional max_timestamp_bound for bounded staleness reads

This commit adds a `max_timestamp_bound` field to the BatchRequest
header. This field places an upper bound on the dynamically chosen
timestamp during a bounded staleness read. If the field is set and a
resolved timestamp over the request's key span is computed to be greater
than the maximum timestamp bound, the batch timestamp will be set to the
maximum timestamp bound instead, then the batch will be evaluated at this
timestamp, and it will be recorded in the BatchResponse Header.

This will be used by SQL to ensure that once a schema version is chosen
for a given table, bounded staleness reads on that table do not observe
data past its expiration time. Without this, it would be possible for
these reads to observe key-value state that is incoherent with the table
descriptor in use.

otan committed:
playing around with nathans pr

Release note (<category, see below>): <what> <show> <why>

otan committed:
error codes

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68457)
<!-- Reviewable:end -->

otan committed:
sqlutils: make a RowsToDataDrivenOutput function

This commit introduces a common way of converting rows to data driven
output. This abstracts this away from backup which is using this in the
hopes that it can be used in other data driven tests I plan to
introduce.

Release note: None
